cirq/contrib/quimb/mps_simulator.py

```diff
@@ -231,92 +235,243 @@ def sample(
-
-            S = np.asarray([math.sqrt(x) for x in S])
-
-            nkeep = 0
-            for i in range(S.shape[0]):
-                if S[i] >= S[0] * self.threshold:
-                    nkeep = i + 1
-
-            X = X[:, :, :nkeep]
-            S = np.diag(S[:nkeep])
-            Y = Y[:nkeep, :, :]
-
-            self.M[n] = np.einsum('mis,sn->mni', X, S)
-            self.M[p] = np.einsum('ns,spj->npj', S, Y)
+        if len(op.qubits) == 1:
+            n = self.qubit_map[op.qubits[0]]
+
+            old_n = self.i_str(n)
+            new_n = 'new_' + old_n
+
+            U = qtn.Tensor(U, inds=(new_n, old_n))
+            self.M[n] = (U @ self.M[n]).reindex({new_n: old_n})
+        elif len(op.qubits) == 2:
+            self.num_svd_splits += 1
+
+            n, p = [self.qubit_map[qubit] for qubit in op.qubits]
+
+            old_n = self.i_str(n)
+            old_p = self.i_str(p)
+            new_n = 'new_' + old_n
+            new_p = 'new_' + old_p
+
+            U = qtn.Tensor(U, inds=(new_n, new_p, old_n, old_p))
+
+            # This is the index on which we do the joining. We need to add it iff it's the first
+            # time that we do the joining for that specific pair.
+            mu_ind = self.mu_str(n, p)
+            if mu_ind not in self.M[n].inds:
+                self.M[n].new_ind(mu_ind)
+            if mu_ind not in self.M[p].inds:
+                self.M[p].new_ind(mu_ind)
+
+            T = U @ self.M[n] @ self.M[p]
+
+            left_inds = tuple(set(T.inds) & set(self.M[n].inds)) + (new_n,)
+            X, Y = T.split(
+                left_inds,
+                cutoff=self.rsum2_cutoff,
+                cutoff_mode='rsum2',
+                get='tensors',
+                absorb='both',
+                bond_ind=mu_ind,
+            )
+
+            self.M[n] = X.reindex({new_n: old_n})
+            self.M[p] = Y.reindex({new_p: old_p})
         else:
+            # NOTE(tonybruguier): There could be a way to handle higher orders. I think this could
+            # involve HOSVDs:
+            # https://en.wikipedia.org/wiki/Higher-order_singular_value_decomposition
+            #
+            # TODO(tonybruguier): Evaluate whether it's even useful to implement and learn more
+            # about HOSVDs.
             raise ValueError('Can only handle 1 and 2 qubit operations')
 
+    def estimation_stats(self):
+        num_coefs_used = sum([Mi.data.size for Mi in self.M])
+        memory_bytes = sum([Mi.data.nbytes for Mi in self.M])
+
+        # The computation below is done for numerical stability, instead of directly using the
+        # formula:
+        # estimated_fidelity = (1 - self.rsum2_cutoff) ** self.num_svd_splits
+        estimated_fidelity = 1.0 + np.expm1(np.log1p(-self.rsum2_cutoff) * self.num_svd_splits)
+        estimated_fidelity = round(estimated_fidelity, ndigits=3)
+
+        return {
+            "num_coefs_used": num_coefs_used,
+            "memory_bytes": memory_bytes,
```

Comment:
check out https://github.com/quantumlib/Cirq/blob/master/cirq/contrib/quimb/state_vector.py#L171 where we just let quimb do the contraction and it can tell us the max RAM throughout the contraction ordering

Reply:
Nice! That deleted a lot of code, including some that you had comments one. Thank you for showing me this! Done.